src/geometry/ArborX_Point.hpp

```diff
@@ -35,11 +37,32 @@ struct Point
   Coordinate _coords[DIM] = {};
 };
 
-template <typename... T>
-KOKKOS_FUNCTION Point(T...)
-    -> Point<sizeof...(T), std::conditional_t<
-                               (... || std::is_same_v<std::decay_t<T>, double>),
-                               double, float>>;
+// If all inputs are integral, convert to double. Otherwise, all inputs must be
+// the same type.
+// Note: converting integral types to double may introduce errors for 8+ byte
+// sized integers.k
+template <typename T, typename... Ts>
+#if KOKKOS_VERSION >= 40400
+KOKKOS_DEDUCTION_GUIDE
+#else
+KOKKOS_FUNCTION
+#endif
+Point(T, Ts...)
+    -> Point<sizeof...(Ts) + 1,
+             std::enable_if_t<std::is_floating_point_v<T> &&
+                                  (... && std::is_floating_point_v<Ts>)&&(
+                                      ... &&std::is_same_v<std::decay_t<T>,
+                                                           std::decay_t<Ts>>),
+                              T>>;
```

Comment:
```suggestion
             std::enable_if_t<std::conjunction_v<std::is_same<T, Ts>...>, T>>;
```
seems to do the job just fine, see https://godbolt.org/z/hMM69qGPr. The error messages are better when the decution guide doesn't constraint to floating point types bu the class has a corresponding `static_assert`.

Reply:
> The error messages are better when the dedution guide doesn't constraint to floating point types but the class has a corresponding `static_assert`.

I don't think it's a good idea to put that static assert in the class. If a user really points to use `Point<D,int>`, we should let them do that.

Reply:
Drop the decay and if Ts are the same as T and T is a floating point you don't need to check that Ts are floating points

Reply:
Done.

Reply:
> I don't think it's a good idea to put that static assert in the class. If a user really points to use `Point<D,int>`, we should let them do that.

Why would we then restrict the deduction guide at all?

Reply:
> Why would we then restrict the deduction guide at all?

I think that `Point{0,0,0}` resulting in `Point<3, int>` may not be what most user expect. I'd rather have them explicitly write `Point<3, int>{0, 0, 0}`.